### ISSUE TITLE
DDLDatabase fails with ParserException on JSON field

### ISSUE THREAD

**Issue by rouzwawi** [jOOQ/jOOQ#7235]
### Expected behavior and actual behavior:
I'm trying to generate java classes for a MySql table containing a `JSON` column. This works with the `org.jooq.util.mysql.MySQLDatabase` generator, but fails with `org.jooq.util.ddl.DDLDatabase`.

Error output is:

```
[ERROR] An exception occurred while parsing script source : src/main/resources/db-schema.sql. Please report this error to https://github.com/jOOQ/jOOQ/issues/new
org.jooq.impl.ParserException: C42000_NO_SUBCLASS: [4:8] y
(
  id   INT AUTO_INCREMENT PRIMARY KEY,
  test [*]JSON NULL,

  CONSTRAINT config_policy_id_uindex
  UNIQUE (id)
);



	at org.jooq.impl.ParserImpl$ParserContext.unexpectedToken(ParserImpl.java:6727)
	at org.jooq.impl.ParserImpl.parseDataType(ParserImpl.java:5698)
	at org.jooq.impl.ParserImpl.parseCreateTable(ParserImpl.java:1544)
	at org.jooq.impl.ParserImpl.parseCreate(ParserImpl.java:1295)
	at org.jooq.impl.ParserImpl.parseQuery(ParserImpl.java:531)
	at org.jooq.impl.ParserImpl.parse(ParserImpl.java:380)
	at org.jooq.impl.ParserImpl.parse(ParserImpl.java:367)
	at org.jooq.util.ddl.DDLDatabase.create0(DDLDatabase.java:129)
	at org.jooq.util.AbstractDatabase.create(AbstractDatabase.java:237)
	at org.jooq.util.AbstractDatabase.create(AbstractDatabase.java:226)
	at org.jooq.util.AbstractDatabase.getDialect(AbstractDatabase.java:208)
	at org.jooq.util.AbstractGenerator.logDatabaseParameters(AbstractGenerator.java:131)
	at org.jooq.util.JavaGenerator.generate(JavaGenerator.java:271)
	at org.jooq.util.GenerationTool.run(GenerationTool.java:642)
	at org.jooq.util.GenerationTool.generate(GenerationTool.java:203)
	at org.jooq.util.maven.Plugin.execute(Plugin.java:192)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:309)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:194)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:107)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:993)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:345)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:191)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
```

### Steps to reproduce the problem:
Set up a JOOQ config file with

```xml
<generator>
    <database>
      <name>org.jooq.util.ddl.DDLDatabase</name>
      <properties>
        <property>
          <key>scripts</key>
          <value>src/main/resources/db-schema.sql</value>
        </property>
      </properties>
  </database>
</generator>
```

`src/main/resources/db-schema.sql`

```sql
CREATE TABLE config_policy
(
  id   INT AUTO_INCREMENT PRIMARY KEY,
  test JSON NULL,

  CONSTRAINT config_policy_id_uindex
  UNIQUE (id)
);
```

Run plugin with config

```xml
<plugin>
  <groupId>org.jooq</groupId>
  <artifactId>jooq-codegen-maven</artifactId>
  <version>3.10.5</version>
  <executions>
    <!-- Do not attach -->
  </executions>
  <dependencies>
    <dependency>
      <groupId>org.jooq</groupId>
      <artifactId>jooq-meta-extensions</artifactId>
      <version>3.10.5</version>
    </dependency>
  </dependencies>
  <configuration>
    <configurationFile>jooq-config.xml</configurationFile>
  </configuration>
</plugin>
```

Run the `generate` goal from command line

```
mvn org.jooq:jooq-codegen-maven:generate
```

### Versions:

- jOOQ: 3.10.5
- Java: 1.8.0_161
- Database (include vendor): MySql (But using DDLDatabase here)
- OS: Mac OSX

**Comment by lukaseder** [comment]
Thank you very much for your detailed report. Indeed, the JSON data type is something that would be very nice to support natively in jOOQ.

**Comment by lukaseder** [comment]
We'll add support for the `JSON` data type in jOOQ 3.11: #7242. Being able to parse it will be a simple task, then.

**Comment by rouzwawi** [comment]
@lukaseder Thanks for the quick reply. Will keep an eye out for 3.11.

**Comment by lukaseder** [comment]
This is fixed in jOOQ 3.12 through #7776

**Comment by lukaseder** [comment]
See also: 

- #7242 support for this type as `SQLDataType`
- #8951 support for translating this type
